elixir/apps/domain/lib/domain/policies/condition/evaluator.ex

```diff
@@ -6,113 +6,183 @@ defmodule Domain.Policies.Condition.Evaluator do
       {:ok, inet} = Domain.Types.INET.cast(cidr)
       cidr = %{inet | netmask: inet.netmask || Domain.Types.CIDR.max_netmask(inet)}
-      Domain.Types.CIDR.contains?(cidr, client.last_seen_remote_ip)
+
+      if Domain.Types.CIDR.contains?(cidr, client.last_seen_remote_ip) do
+        {:halt, {:ok, nil}}
+      else
+        {:cont, :error}
+      end
     end)
   end
 
-  def conforms?(
+  def fetch_conformation_expiration(
         %Condition{property: :remote_ip, operator: :is_not_in_cidr, values: values},
         %Clients.Client{} = client
       ) do
-    Enum.all?(values, fn cidr ->
+    Enum.reduce_while(values, {:ok, nil}, fn cidr, {:ok, nil} ->
       {:ok, inet} = Domain.Types.INET.cast(cidr)
       cidr = %{inet | netmask: inet.netmask || Domain.Types.CIDR.max_netmask(inet)}
-      not Domain.Types.CIDR.contains?(cidr, client.last_seen_remote_ip)
+
+      if Domain.Types.CIDR.contains?(cidr, client.last_seen_remote_ip) do
+        {:halt, :error}
+      else
+        {:cont, {:ok, nil}}
+      end
     end)
   end
 
-  def conforms?(
+  def fetch_conformation_expiration(
         %Condition{property: :provider_id, operator: :is_in, values: values},
         %Clients.Client{} = client
       ) do
     client = Repo.preload(client, :identity)
-    client.identity.provider_id in values
+
+    if client.identity.provider_id in values do
+      {:ok, nil}
+    else
+      :error
+    end
   end
 
-  def conforms?(
+  def fetch_conformation_expiration(
         %Condition{property: :provider_id, operator: :is_not_in, values: values},
         %Clients.Client{} = client
       ) do
     client = Repo.preload(client, :identity)
-    client.identity.provider_id not in values
+
+    if client.identity.provider_id in values do
+      :error
+    else
+      {:ok, nil}
+    end
   end
 
-  def conforms?(
+  def fetch_conformation_expiration(
         %Condition{
           property: :current_utc_datetime,
           operator: :is_in_day_of_week_time_ranges,
           values: values
         },
         %Clients.Client{}
       ) do
-    datetime_in_day_of_the_week_time_ranges?(DateTime.utc_now(), values)
+    case find_day_of_the_week_time_range(values, DateTime.utc_now()) do
+      nil -> :error
+      expires_at -> {:ok, expires_at}
+    end
   end
 
-  def datetime_in_day_of_the_week_time_ranges?(datetime, dow_time_ranges) do
+  def find_day_of_the_week_time_range(dow_time_ranges, datetime) do
     dow_time_ranges
     |> parse_days_of_week_time_ranges()
     |> case do
       {:ok, dow_time_ranges} ->
-        Enum.any?(dow_time_ranges, fn {day, time_ranges} ->
+        dow_time_ranges
+        |> Enum.find_value(fn {day, time_ranges} ->
+          time_ranges = merge_joint_time_ranges(time_ranges)
           datetime_in_time_ranges?(datetime, day, time_ranges)
         end)
 
       {:error, _reason} ->
-        false
+        nil
     end
   end
 
+  # Merge ranges, eg. 4-11,11-22 = 4-22
+  defp merge_joint_time_ranges(time_ranges) do
```

Comment:
Would it be a good idea to mark this function public and have a few unit tests just for it?

Reply:
@jamilbk we have a few cases that test this indirectly but I'll add a few unit tests directly for it too